### ISSUE TITLE
superfluous characters in text/symbols.py ?

### ISSUE THREAD

**Issue by NightFantomJ2** [keithito/tacotron#108]
Since all cleaners include a 
`  text = lowercase(text)`
line, aren't the uppercase characters in symbols.py redundant?

As far as I can tell any input is passed through the cleaners before being learned by the network, so those uppercase input neurons will never get a nonzero input, so the network will never learn for them.

Or did I miss something?

Semi-related: is there a way to continue from a checkpoint with a different amount of symbols?

**Comment by keithito** [comment]
Yes, that's correct. However, it would be an interesting experiment to try not lowercasing the text. There's probably some useful signal in there with proper nouns, acronyms, etc.

Except for using a little bit of memory to store the embeddings, there's no penalty to having extra symbols in there. As you mention, they'll never get backpropped into, so they'll just remain at their initial values and will not play a role in the model.

I don't know of a way to continue from a checkpoint after adding symbols. However, it may be possible to tell tensorflow to load the checkpoint and initialize the missing weights. I'm not sure.

**Comment by NightFantomJ2** [comment]
> However, it would be an interesting experiment to try not lowercasing the text. There's probably some useful signal in there with proper nouns, acronyms, etc.

I agree, I fear how much longer it would take training with capital letters too though.

>Except for using a little bit of memory to store the embeddings, there's no penalty to having extra symbols in there. As you mention, they'll never get backpropped into, so they'll just remain at their initial values and will not play a role in the model.

I see, do you have any insights in the memory requirements per character? I'm thinking if superfluous characters aren't *that* bad, it might be worth a shot to train on full ascii or even full utf-8. I'll need to look a lot deeper under the hood of both tensorflow and your tacotron implementation to understand what's going on 😄.

> I don't know of a way to continue from a checkpoint after adding symbols. However, it may be possible to tell tensorflow to load the checkpoint and initialize the missing weights. I'm not sure.

The reason I'm asking is because I'm training different configurations in parallel and the one making the best progress for now is the one starting from your English 870k'th iteration, then training with my Dutch input. I'm assuming it's because English and Dutch aren't that different in pronunciation that it's able to keep the learned alignment and train the Dutch voice on top of that.

But the issue with that is that I seem restricted to the initial characters you used, so for example I can't re-use it with a language with more characters.

Thanks once again!

**Comment by keithito** [comment]
> I see, do you have any insights in the memory requirements per character? I'm thinking if superfluous characters aren't that bad, it might be worth a shot to train on full ascii or even full utf-8. I'll need to look a lot deeper under the hood of both tensorflow and your tacotron implementation to understand what's going on 😄.

I think it should be relatively small. For each additional character, you have a 256D embedding, which should be 1KB for the embedding itself, and then maybe another 1KB for the gradient. I think you can use the [Tensorflow Profiler](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/core/profiler) to know for sure.

> But the issue with that is that I seem restricted to the initial characters you used, so for example I can't re-use it with a language with more characters.

Ah, that's a good point! I wonder if it would be possible to write a tool that would read the checkpoint, resize the embedding table, and then save a new checkpoint. I don't know enough about the internals of how the Saver works to know how hard that would be.
